google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb

```diff
@@ -0,0 +1,566 @@
+        #   listener.stopped? #=> true
+        #
+        def stopped?
+          synchronize { @stopped }
+        end
+
+        private
+
+        def send_callback query_snp
+          @callback.call query_snp
+        end
+
+        def start_listening!
+          # create new background thread to handle the stream's enumerator
+          @background_thread = Thread.new { background_run }
+        end
+
+        # @private
+        class RestartStream < StandardError; end
+
+        # rubocop:disable all
+
+        def background_run
+          # Don't allow a stream to restart if already stopped
+          return if synchronize { @stopped }
+
+          @backoff ||= { max: 5, current: 0, delay: 1.0, mod: 1.3 }
+
+          # Reuse inventory if one already exists
+          # Even though this uses an @var, no need to synchronize
+          @inventory ||= Inventory.new(synchronize { @query })
+          @inventory.reset
+
+          # Send stop if already running
+          synchronize do
+            @request_queue.push self if @request_queue
+          end
+
+          # Customize the provided initial listen request
+          init_listen_req = Google::Firestore::V1beta1::ListenRequest.new(
+            database: synchronize { @query.client.path },
+            add_target: Google::Firestore::V1beta1::Target.new(
+              query: Google::Firestore::V1beta1::Target::QueryTarget.new(
+                parent: synchronize { @query.parent_path },
+                structured_query: synchronize { @query.query }
+              ),
+              resume_token: String(@inventory.resume_token),
+              read_time: Convert.time_to_timestamp(@inventory.read_time),
+              target_id: 0x42
+            )
+          )
+
+          # Always create a new enum queue
+          synchronize do
+            @request_queue = EnumeratorQueue.new self
+            @request_queue.push init_listen_req
+          end
+
+          # Not an @var, we get a new enum each time
+          enum = synchronize do
+            @query.client.service.listen @request_queue.each
+          end
+
+          loop do
+
+            # Break loop, close thread if stopped
+            break if synchronize { @stopped }
+
+            begin
+              # Cannot syncronize the enumerator, causes deadlock
+              response = enum.next
+
+              case response.response_type
+              when :target_change
+                case response.target_change.target_change_type
+                when :NO_CHANGE
+                  # No change has occurred. Used only to send an updated
+                  # +resume_token+.
+
+                  @inventory.persist(
+                    response.target_change.resume_token,
+                    Convert.timestamp_to_time(
+                      response.target_change.read_time
+                    )
+                  )
+
+                  if @inventory.current? && @inventory.changes?
```

Comment:
As I mentioned before, I separated the implementations because it was too difficult to make changes to. I consider this PR to still be in the red phase of red/green/refactor. Once the PR is approved then we can try refactoring the implementation using the test coverage we have built. But for now I am leaving these implementations separate.